### ISSUE TITLE
Getting rid of '<unknown description>' from SeqRecord object

### ISSUE THREAD

**Issue by mmokrejs** [biopython/biopython#831]
Hi,
  I just need to create an object to get a reverse-complement of it and write out a FASTA and FASTQ file. I have the following function:

```
from Bio.SeqRecord import SeqRecord
from Bio.Seq import Seq, reverse_complement
from Bio.Alphabet import generic_dna

def create_SeqObj_from_sequence_and_quals(myid, description, sequence, quals, alphabet=generic_dna):
    if sequence and myid:
        if description:
            _record = SeqRecord(Seq(sequence, alphabet), id=myid, description=description)
        else:
            _record = SeqRecord(Seq(sequence, alphabet), id=myid)
        if quals:
            _record.letter_annotations["phred_quality"]=quals
        return _record
    else:
        return None
```

Unfortunately, this does not work as "expected" when I pass the object to `SeqIO.write(_record, out_fh_seq, 'fastq')` it gives me aside the FASTQ ID name a `'<unknown description>'` in the description line. This is not nice.

Here is what the object contains:
```
>>> _record=create_SeqObj_from_sequence_and_quals('A21233', '', 'atgc', [40, 40, 40, 40])
>>> _record
SeqRecord(seq=Seq('atgc', DNAAlphabet()), id='A21233', name='<unknown name>', description='<unknown description>', dbxrefs=[])
>>> 
```

So, could either SeqIO.write() ignore the _record.description and _record.name stub values, or even better, could SeqRecord keep the attributes just unset?

I have biopython-1.66.

**Comment by peterjc** [comment]
Short term, since you don't like the defaults, set the values you want explicitly:

```
        if description:
            _record = SeqRecord(Seq(sequence, alphabet), id=myid, description=description, name="")
        else:
            _record = SeqRecord(Seq(sequence, alphabet), id=myid, description="", name="")
```

Changing the defaults makes me nervous given how long the current values have been in place - what would you use instead? Empty strings? ``None``?

**Comment by mmokrejs** [comment]
I would be happy if `''` and `None` gave same result (no placeholder value output in FASTA/Q files through SeqIO).


You are right, forcing `_record.description=''` works fine for my purpose. I was lazy to check what SeqIO does but now I can confirm it outputs just the FASTA ID alone on one line, which is exactly what I wanted:

```
$ cat /tmp/test.fastq 
@A21233
atgc
+
IIII
$
```

```
$ od -c /tmp/test.fastq 
0000000   @   A   2   1   2   3   3  \n   a   t   g   c  \n   +  \n   I
0000020   I   I   I  \n
0000024
$
```
Thank you for the reminder, I will use this workaround, it is OK for my purpose.

BTW, updating the last section of [http://biopython.org/wiki/SeqRecord] to include manually defined QUAL values would be a good idea.

**Comment by peterjc** [comment]
Currently the last example in http://biopython.org/wiki/SeqRecord on creating a `SeqRecord` is a protein sequence, so adding quality scores would be a bit odd.

Putting ``None`` as a default value is probably going to break too many existing scripts which expect a string and would suddenly get exceptions. Putting ``''`` (empty string) as a default might be safer.

**Comment by mmokrejs** [comment]
It is true the example is about a protein sequence but I still think changing it to a DNA sequence with QUALs would be more educative (most of the page is about nucleotide sequences although I did not read it that thoroughly). Evidently biopython lacks an example on that, and automated tests, otherwise the `'<unknown description>'` string and `'<unknown name>'` would be realized much sooner in output files.

**Comment by peterjc** [comment]
If you create a ``SeqRecord`` object using these strings, then Biopython will write these strings out if the output format supports relevant fields to put them in - its not a exactly a bug, its a design choice which has drawbacks. Personally I think empty strings or ``None`` would have been be a better default (a choice made right back at the start of Biopython https://github.com/biopython/biopython/commit/f726249141aa42afbfabab72157fe0f7d820b548 many years before my time), and if we were to make the change now, my guess is using empty strings would have least side effects.

Do you want to try that out and see what breaks?

**Comment by mmokrejs** [comment]
I looked for "biopython 'unknown description'" in Google and so far the only related matches are: 
[(http://www.agapow.net/science/computational-biology/biofoo/hitchhikers-guide-to-biopython-seqrecords/)]
[(http://biopython.org/wiki/Ace_contig_class)]
[(http://ccbbatut.github.io/Biocomputing_Spring2016/python5_biopython/)]
[(http://mailman.open-bio.org/pipermail/biopython-dev/2012-December/019321.html)]

The issue is also mentioned in [(http://biorpy.blogspot.cz/2015/03/bpy7-writing-multi-record-sequences.html)].

As I said, I am fine with empty string as the default, and I agree None may be asking for a trouble through __repr__. I think the thread with HSP objects and inherited values is of possibly the main concern here.

**Comment by peterjc** [comment]
Please clarify which HSP inheritance issue you're referring to. The only code I could find in ``Bio.SearchIO`` to create a ``SeqRecord`` does so from an HSP, but it does set the name and description explicitly: https://github.com/biopython/biopython/blob/master/Bio/SearchIO/_model/hsp.py#L832

**Comment by mmokrejs** [comment]
I mean what Bow said in [(http://mailman.open-bio.org/pipermail/biopython-dev/2012-December/019321.html)]. Could be he improved SearchIO since that time to prevent such issues. I could quote the whole message but here is at least the key part IMHO:

> '<unknown_description>' is the default value for any description
> attribute (be it in the QueryResult object, or in the
> HSPFragment.hit_description). The error you're seeing is because the
> hit description is being accessed through the hit object
> (hit.description) and the cascading property getter checks first whether all
> HSP contains the same `hit_description` attribute value. It'll only
> return the value if all HSPFragment.hit_description values are equal.
> Otherwise, it'll raise the error you're seeing here.

**Comment by peterjc** [comment]
Ah - I thought we were still talking about the ``SeqRecord`` defaults. Thanks for the clarification.

**Comment by mmokrejs** [comment]
I am sorry, I was trying to get myself some broader glimpse where it is either used or even, if anybody realized this in the past. I think Bow should chim in, possibly the line [(https://github.com/biopython/biopython/blob/master/Bio/SearchIO/_model/hsp.py#L694]) should be kept in sync with the eventual future new default, '' empty string. I do not have a overview of biopython's guts so I am splashing here. I am happy to "test" the change but doubt being productive in this regard. Automated tests will do better than me, and most importantly file parsers will break quickly if they are incompatible.

The comments in [(https://github.com/biopython/biopython/commit/f726249141aa42afbfabab72157fe0f7d820b548)] in `Bio/SeqRecord.py` docstring would encourage me to do the change. ;-)

**Comment by bow** [comment]
Hi @mmokrejs & @peterjc,

The `<unknown description>` default I wrote in SearchIO was with the idea to keep blank description values with `SeqIO` actually.

As you've probably seen, SearchIO uses regular SeqRecord objects to represent the actual sequence (if any). This of course lies very deep in the object hierarchy (it's within `HSPFragment`, which is contained by `HSP`, which in turn is contained by `Hit`, which is finally contained by `QueryResult`).

IIRC, I wanted to minimize surprises, so I went with what `SeqIO` had as the default empty value for sequences with no descriptions. If we do change the `SeqIO` default, the `SearchIO` default should follow as well (which I have no problem with, by the way). The way I see it, `SearchIO` is a client library of `SeqIO` and should try to follow its convention as much as possible.

**Comment by peterjc** [comment]
Someone could try using the empty string as the default (or even ``None``) on a branch and see how much of the test suite breaks. Most of the the parsers should be fine because they explicitly set these values to something - but end user code is harder to predict.

**Comment by mmokrejs** [comment]
Interestingly, I hit this problem again when reverse-complementing a sequence object in biopython 1.81.

```
$ head -n 4 test.txt 
>A00808:1538:HWL7GDRX3:2:2101:7491:1125 minus
AGAAAGTACTACTAATCTGTATGGTAGGTGACCAACACCATAAGTGGGTCGGAAACCATATGATTGTAAAGGAGAGTAACAATTAGGACCTGCAACACAAATACAAGGTTTGATACCGGCCAGATAGATATCAGTTGAAATATATCTCTCAAAAGGTTTCAGTTTAGACTTCCTAAACAATATATACCAGTAATCATAAATACCACTATGCTTAGAATCAAGCATGTTAGAATTCCAAGCTATATCGCAGCCTTTTAAATCATCTGGTAATTTTTAATTATAATCAGCAATATTTCCAGTTTGCCCTGGAGCGATTTGGCTGACTTCATTACCTTTAATTACAAATGAATCTGCATAGACATGAGTAAAGCAGTGATCATTTATTTTAGTAGGAGACACTCC
>A00808:1538:HWL7GDRX3:2:2101:21296:1204 plus
GGAGTGTCTCCTACTAAATTAAATGATCTCTGCTTTACTAATGTCTATGCAGATTCATTTGTAATTAAAGGTAATGAAGTCAGCCAAATCGCTCCAGGGCAAACTGGAAATATTGCTGATTATAATTATAAATTACCAGATGATTTTACAGGCTGCGTTATAGCTTGGAATTCTAACAAGCTTGATTCTAAGCATAGTGGTAATTATGATTACTGGTATAGATTGTTTAGGAAGTCTAAACTCAAACCTTTTGAGAGAGATATTTCAACTGAAATCTATCAGGCCGGTAACAAACCTTGTAAAGG---TAAAGGTCCTAATTGTTACTTTCCTTTACAATCATATGGTTTCCGACCCACTTATGGTGTTGGTCACCAACCATACAGAGTAGTAGTACTTTCT
$ head -n 4 test.txt | python reversecomplement_reads_on_minus.py --infile -
><unknown id> <unknown description>
GGAGTGTCTCCTACTAAAATAAATGATCACTGCTTTACTCATGTCTATGCAGATTCATTTGTAATTAAAGGTAATGAAGTCAGCCAAATCGCTCCAGGGCAAACTGGAAATATTGCTGATTATAATTAAAAATTACCAGATGATTTAAAAGGCTGCGATATAGCTTGGAATTCTAACATGCTTGATTCTAAGCATAGTGGTATTTATGATTACTGGTATATATTGTTTAGGAAGTCTAAACTGAAACCTTTTGAGAGATATATTTCAACTGATATCTATCTGGCCGGTATCAAACCTTGTATTTGTGTTGCAGGTCCTAATTGTTACTCTCCTTTACAATCATATGGTTTCCGACCCACTTATGGTGTTGGTCACCTACCATACAGATTAGTAGTACTTTCT
>A00808:1538:HWL7GDRX3:2:2101:21296:1204 plus
GGAGTGTCTCCTACTAAATTAAATGATCTCTGCTTTACTAATGTCTATGCAGATTCATTTGTAATTAAAGGTAATGAAGTCAGCCAAATCGCTCCAGGGCAAACTGGAAATATTGCTGATTATAATTATAAATTACCAGATGATTTTACAGGCTGCGTTATAGCTTGGAATTCTAACAAGCTTGATTCTAAGCATAGTGGTAATTATGATTACTGGTATAGATTGTTTAGGAAGTCTAAACTCAAACCTTTTGAGAGAGATATTTCAACTGAAATCTATCAGGCCGGTAACAAACCTTGTAAAGG---TAAAGGTCCTAATTGTTACTTTCCTTTACAATCATATGGTTTCCGACCCACTTATGGTGTTGGTCACCAACCATACAGAGTAGTAGTACTTTCT
$
```

```
$ cat reversecomplement_reads_on_minus.py
#! /usr/bin/env python

import sys, os
import tempfile
from Bio import SeqIO
from Bio.Seq import Seq, reverse_complement
from Bio.SeqRecord import SeqRecord

from optparse import OptionParser

myparser = OptionParser(version="%s version %s" % ('%prog', version))
myparser.add_option("--infile", action="store", type="string", dest="infile", default='',
    help="Input FASTA file with word plus or minus as the second word of the header, use minus (dash) for stdin")
myparser.add_option("--outfile", action="store", type="string", dest="outfile", default='',
    help="Output filename with words plus and minus removed from FASTA header lines, use minus (dash) for stdout")
myparser.add_option("--infileformat", action="store", type="string", dest="infileformat", default='fasta',
    help="Input file format (Default: fasta)")
myparser.add_option("--outfileformat", action="store", type="string", dest="outfileformat", default='fasta-2line',
    help="Output file format (Default: fasta)")
(myoptions, myargs) = myparser.parse_args()


def parse_input(infile, infileformat):
    for record in SeqIO.parse(infile, infileformat):
        #print(dir(record))
        #print(record.description)
        _fasta_header_items = record.description.split()
        if len(_fasta_header_items) > 1:
            if _fasta_header_items[1] == 'minus':
                yield(record.reverse_complement())
            else:
                yield(record)

if __name__ == "__main__":
    if myoptions.infile == '-':
        _infile = sys.stdin
    elif os.path.exists(myoptions.infile):
        _infile = myoptions.infile
    else:
        raise RuntimeError("File %s does not exist" % myoptions.infile)

    if not myoptions.outfile:
        _outfile = sys.stdout
    elif myoptions.outfile == '-':
        _outfile = sys.stdout
    else:
        _outfile = myoptions.outfile

    SeqIO.write(parse_input(_infile, myoptions.infileformat), _outfile, myoptions.outfileformat)
$
```

**Comment by mmokrejs** [comment]
@peterjc Thank you for the great tips and reminder for the docstring. I agree `_rc` would make sense in common situations. I was just surprised that with the defaults I tripped into this. I think `record.reverse_complement()` might better append `_rc` by default than just letting slip through the historical object defaults discussed above.

BTW: I am merging fwd and rev pair members together before this step using `fastp` so in this case it does not make sense to keep orientation or the two halves. Surprisingly I did not find a tool to merge a read pair using alignment to a reference (I want to avoid overlap-based approaches) a resolve conflicts between paired read respecting the nucleotide with higher QUAL. If reads do not overlapp, fill-in the gap with Ns or alternatively, copy over the flank from the reference.
